BrainPortal/app/controllers/data_providers_controller.rb

```diff
@@ -399,263 +399,328 @@ def browse
+      ! current_user.available_groups.map(&:id).include?(group_id)
+    )
+
+    # Register the given userfiles in background.
+    userfiles = userfiles_from_basenames(@provider, @as_user, params[:basenames])
+    userfiles_count = userfiles.count # Avoids a cute race condition
+
+    registered, already_registered = [], []
+    succeeded, failed = [], {}
+
+    CBRAIN.spawn_with_active_records_if(
+      [:html, :js].include?(request.format.to_sym),
+      current_user,
+      "Register files (data_provider: #{@provider.id})"
+    ) do
+      userfiles.keys.shuffle.each do |basename|
+        begin
+          # Is the file already registered?
+          if userfiles[basename].present?
+            already_registered << userfiles[basename]
+            (failed["Already registered"] ||= []) << basename
+            next
+          end
 
-    flash[:error]  = ""
-    flash[:notice] = ""
+          # Determine the filetype of the new file
+          subtype = filetypes[basename] || "SingleFile"
+          unless valid_types.include?(subtype)
+            (failed["Unknown type #{subtype}"] ||= []) << basename
+            next
+          end
 
-    legal_subtypes = Userfile.descendants.map(&:name).index_by { |x| x }
+          # Create the new userfile
+          userfile = subtype.constantize.new(
+            :name             => basename,
+            :user_id          => @as_user.id,
+            :group_id         => group_id,
+            :data_provider_id => @provider.id
+          )
+
+          # And save it
+          if userfile.save
+            userfile.addlog_context(self, "Registered on DataProvider '#{@provider.name}' as '#{userfile.name}' by #{current_user.login}.")
+            registered << (userfiles[basename] = userfile)
+            succeeded << basename
+          else
+            (failed["Unspecified error"] ||= []) << basename
+          end
 
-    # NOTE: next paragraph for initializing registered_files is also in browse() action
-    registered_files = Userfile.where( :data_provider_id => @provider.id )
-    # On data providers where files are stored in a per user subdir, we limit our
-    # search of what's registered to only those belonging to @as_user
-    registered_files = registered_files.where( :user_id => @as_user.id ) if ! @provider.allow_file_owner_change?
-    registered_files = registered_files.all.index_by(&:name)
+        rescue => e
+          (failed[e.message] ||= []) << basename
+        end
+      end
 
-    basenames.shuffle.each do |basename|
+      # Remind the user if browsing as another user
+      flash[:notice] += "Important note! Since you were browsing as user '#{@as_user.login}', the files were registered as belonging to that user instead of you!\n" if
+        succeeded.present? && @as_user != current_user
 
-      # Unregister files
+      # If files actually got registered, clear the browsing cache
+      clear_browse_provider_local_cache_file(@as_user, @provider) if
```

Comment:
Didn't we already have this kind of situation before anyway when a user is browsing while someone else is registering/deleting/unregistering files? Assuming the browsing actions automatically regenerate the cache, the only issue I can see is if both actions run exactly at the same time, a rather unlikely (famous last words) situation.